// @omni-door/utils@1.3.2/lib/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var exec_1 = __importDefault(require("./exec"));
var logPrefix_1 = __importStar(require("./logPrefix"));
var logger_1 = require("./logger");
var dependencies_1 = require("./dependencies");
var pkgName_1 = __importDefault(require("./pkgName"));
var nodeVersion_1 = __importDefault(require("./nodeVersion"));
var npmVersion_1 = __importStar(require("./npmVersion"));
var outputFile_1 = __importDefault(require("./outputFile"));
var spinner_1 = __importDefault(require("./spinner"));
var tplEngine_1 = require("./tplEngine");
var requireCwd_1 = __importDefault(require("./requireCwd"));
var gitHandlers_1 = require("./gitHandlers");
var updateNotifier_1 = __importDefault(require("./updateNotifier"));
var exec_2 = require("./exec");
Object.defineProperty(exports, "exec", { enumerable: true, get: function () { return exec_2.default; } });
var logPrefix_2 = require("./logPrefix");
Object.defineProperty(exports, "logPrefix", { enumerable: true, get: function () { return logPrefix_2.default; } });
Object.defineProperty(exports, "setLogo", { enumerable: true, get: function () { return logPrefix_2.setLogo; } });
Object.defineProperty(exports, "setBrand", { enumerable: true, get: function () { return logPrefix_2.setBrand; } });
Object.defineProperty(exports, "getLogo", { enumerable: true, get: function () { return logPrefix_2.getLogo; } });
Object.defineProperty(exports, "getBrand", { enumerable: true, get: function () { return logPrefix_2.getBrand; } });
var logger_2 = require("./logger");
Object.defineProperty(exports, "logDetail", { enumerable: true, get: function () { return logger_2.logDetail; } });
Object.defineProperty(exports, "logEmph", { enumerable: true, get: function () { return logger_2.logEmph; } });
Object.defineProperty(exports, "logErr", { enumerable: true, get: function () { return logger_2.logErr; } });
Object.defineProperty(exports, "logInfo", { enumerable: true, get: function () { return logger_2.logInfo; } });
Object.defineProperty(exports, "logSuc", { enumerable: true, get: function () { return logger_2.logSuc; } });
Object.defineProperty(exports, "logWarn", { enumerable: true, get: function () { return logger_2.logWarn; } });
Object.defineProperty(exports, "logTime", { enumerable: true, get: function () { return logger_2.logTime; } });
Object.defineProperty(exports, "logCongrat", { enumerable: true, get: function () { return logger_2.logCongrat; } });
Object.defineProperty(exports, "underline", { enumerable: true, get: function () { return logger_2.underline; } });
Object.defineProperty(exports, "italic", { enumerable: true, get: function () { return logger_2.italic; } });
var dependencies_2 = require("./dependencies");
Object.defineProperty(exports, "getDependency", { enumerable: true, get: function () { return dependencies_2.getDependency; } });
Object.defineProperty(exports, "arr2str", { enumerable: true, get: function () { return dependencies_2.arr2str; } });
Object.defineProperty(exports, "intersection", { enumerable: true, get: function () { return dependencies_2.intersection; } });
var pkgName_2 = require("./pkgName");
Object.defineProperty(exports, "pkgNameCheck", { enumerable: true, get: function () { return pkgName_2.default; } });
var nodeVersion_2 = require("./nodeVersion");
Object.defineProperty(exports, "nodeVersionCheck", { enumerable: true, get: function () { return nodeVersion_2.default; } });
var npmVersion_2 = require("./npmVersion");
Object.defineProperty(exports, "npmVersionCheck", { enumerable: true, get: function () { return npmVersion_2.default; } });
Object.defineProperty(exports, "getNpmVersion", { enumerable: true, get: function () { return npmVersion_2.getNpmVersion; } });
Object.defineProperty(exports, "getNpmVersions", { enumerable: true, get: function () { return npmVersion_2.getNpmVersions; } });
var outputFile_2 = require("./outputFile");
Object.defineProperty(exports, "outputFile", { enumerable: true, get: function () { return outputFile_2.default; } });
var spinner_2 = require("./spinner");
Object.defineProperty(exports, "spinner", { enumerable: true, get: function () { return spinner_2.default; } });
var tplEngine_2 = require("./tplEngine");
Object.defineProperty(exports, "tplEngineInit", { enumerable: true, get: function () { return tplEngine_2.tplEngineInit; } });
Object.defineProperty(exports, "tplEngineNew", { enumerable: true, get: function () { return tplEngine_2.tplEngineNew; } });
Object.defineProperty(exports, "_typeof", { enumerable: true, get: function () { return tplEngine_2._typeof; } });
var requireCwd_2 = require("./requireCwd");
Object.defineProperty(exports, "requireCwd", { enumerable: true, get: function () { return requireCwd_2.default; } });
var gitHandlers_2 = require("./gitHandlers");
Object.defineProperty(exports, "isInGitRepository", { enumerable: true, get: function () { return gitHandlers_2.isInGitRepository; } });
Object.defineProperty(exports, "tryGitInit", { enumerable: true, get: function () { return gitHandlers_2.tryGitInit; } });
var updateNotifier_2 = require("./updateNotifier");
Object.defineProperty(exports, "updateNotifier", { enumerable: true, get: function () { return updateNotifier_2.default; } });
exports.default = {
    exec: exec_1.default,
    logPrefix: logPrefix_1.default,
    setLogo: logPrefix_1.setLogo,
    setBrand: logPrefix_1.setBrand,
    getLogo: logPrefix_1.getLogo,
    getBrand: logPrefix_1.getBrand,
    logDetail: logger_1.logDetail,
    logEmph: logger_1.logEmph,
    logErr: logger_1.logErr,
    logInfo: logger_1.logInfo,
    logSuc: logger_1.logSuc,
    logWarn: logger_1.logWarn,
    logCongrat: logger_1.logCongrat,
    logTime: logger_1.logTime,
    underline: logger_1.underline,
    italic: logger_1.italic,
    pkgNameCheck: pkgName_1.default,
    nodeVersionCheck: nodeVersion_1.default,
    npmVersionCheck: npmVersion_1.default,
    getNpmVersion: npmVersion_1.getNpmVersion,
    outputFile: outputFile_1.default,
    getDependency: dependencies_1.getDependency,
    arr2str: dependencies_1.arr2str,
    intersection: dependencies_1.intersection,
    spinner: spinner_1.default,
    tplEngineInit: tplEngine_1.tplEngineInit,
    tplEngineNew: tplEngine_1.tplEngineNew,
    _typeof: tplEngine_1._typeof,
    requireCwd: requireCwd_1.default,
    isInGitRepository: gitHandlers_1.isInGitRepository,
    tryGitInit: gitHandlers_1.tryGitInit,
    updateNotifier: updateNotifier_1.default
};
