// @omni-door/utils@1.3.2/lib/logger.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.italic = exports.underline = exports.logTime = exports.logCongrat = exports.logSuc = exports.logEmph = exports.logDetail = exports.logInfo = exports.logWarn = exports.logErr = exports.interactive = void 0;
var chalk_1 = __importDefault(require("chalk"));
var signale_1 = require("signale");
var logPrefix_1 = __importStar(require("./logPrefix"));
var emoji_1 = require("./emoji");
var signale = new signale_1.Signale({
    scope: logPrefix_1.getBrand()
});
exports.interactive = function () { return new signale_1.Signale({
    interactive: true,
    scope: logPrefix_1.getBrand()
}); };
function logErr(err) {
    console.error(chalk_1.default.redBright(logPrefix_1.default(), err, emoji_1.stateMap2Emoji.fail, '\n'));
}
exports.logErr = logErr;
function logWarn(warn) {
    console.warn(chalk_1.default.yellowBright(logPrefix_1.default(), warn, emoji_1.stateMap2Emoji.warn, '\n'));
}
exports.logWarn = logWarn;
function logInfo(info) {
    console.info(chalk_1.default.white(logPrefix_1.default(), info, '\n'));
}
exports.logInfo = logInfo;
function logDetail(detail) {
    console.log(chalk_1.default.gray(logPrefix_1.default(), detail, '\n'));
}
exports.logDetail = logDetail;
function logEmph(info) {
    console.info(chalk_1.default.cyanBright(logPrefix_1.default(), info, emoji_1.stateMap2Emoji.hot, '\n'));
}
exports.logEmph = logEmph;
function logSuc(msg) {
    console.info(chalk_1.default.greenBright(logPrefix_1.default(), chalk_1.default.bold(msg), emoji_1.stateMap2Emoji.succeed, '\n'));
}
exports.logSuc = logSuc;
function logCongrat(msg) {
    console.info(chalk_1.default.magentaBright(logPrefix_1.default(), chalk_1.default.bold(msg), emoji_1.stateMap2Emoji.congratulation, '\n'));
}
exports.logCongrat = logCongrat;
function logTime(msg, isEnd) {
    var method = isEnd ? 'timeEnd' : 'time';
    var newSignale = signale.scope(logPrefix_1.getBrand());
    return newSignale[method](chalk_1.default.gray(logPrefix_1.getLogo() + "  " + msg));
}
exports.logTime = logTime;
function underline(str) {
    return chalk_1.default.bold.underline(str);
}
exports.underline = underline;
function italic(str) {
    return chalk_1.default.bold.italic(str);
}
exports.italic = italic;
