// @omni-door/cli@2.8.2/lib/commands/initial/initial_preset.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.cli_toolkit = exports.cli_components_vue = exports.cli_components_react = exports.cli_ssr_react = exports.cli_entire_vue = exports.cli_standard_vue = exports.cli_basic_vue = exports.cli_pc_react = exports.cli_entire_react = exports.cli_standard_react = exports.cli_basic_react = void 0;
exports.cli_basic_react = {
    project_type: 'spa-react',
    pkgtool: 'pnpm',
    build: 'webpack',
    ts: false,
    testFrame: '',
    eslint: false,
    prettier: false,
    commitlint: false,
    style: 'css',
    layout: 'px',
    stylelint: false
};
exports.cli_standard_react = {
    project_type: 'spa-react',
    pkgtool: 'pnpm',
    build: 'webpack',
    ts: true,
    testFrame: '',
    eslint: true,
    prettier: true,
    commitlint: false,
    style: 'less',
    layout: 'px',
    stylelint: true
};
exports.cli_entire_react = {
    project_type: 'spa-react',
    pkgtool: 'pnpm',
    build: 'webpack',
    ts: true,
    testFrame: 'jest',
    eslint: true,
    prettier: true,
    commitlint: true,
    style: 'all',
    layout: 'px',
    stylelint: true
};
exports.cli_pc_react = {
    project_type: 'spa-react-pc',
    pkgtool: 'pnpm',
    build: 'webpack',
    ts: true,
    testFrame: '',
    eslint: true,
    prettier: true,
    commitlint: true,
    style: 'less',
    stylelint: true
};
exports.cli_basic_vue = {
    project_type: 'spa-vue',
    pkgtool: 'pnpm',
    build: 'webpack',
    ts: false,
    testFrame: '',
    eslint: false,
    prettier: false,
    commitlint: false,
    style: 'css',
    layout: 'px',
    stylelint: false
};
exports.cli_standard_vue = {
    project_type: 'spa-vue',
    pkgtool: 'pnpm',
    build: 'webpack',
    ts: true,
    testFrame: '',
    eslint: true,
    prettier: true,
    commitlint: false,
    style: 'less',
    layout: 'px',
    stylelint: true
};
exports.cli_entire_vue = {
    project_type: 'spa-vue',
    pkgtool: 'pnpm',
    build: 'webpack',
    ts: true,
    testFrame: 'jest',
    eslint: true,
    prettier: true,
    commitlint: true,
    style: 'all',
    layout: 'px',
    stylelint: true
};
exports.cli_ssr_react = {
    project_type: 'ssr-react',
    pkgtool: 'pnpm',
    build: 'next',
    ts: true,
    testFrame: 'jest',
    eslint: true,
    prettier: true,
    commitlint: true,
    style: 'all',
    stylelint: true,
    serverType: 'next'
};
exports.cli_components_react = {
    project_type: 'component-react',
    pkgtool: 'yarn',
    build: 'tsc',
    ts: true,
    testFrame: 'jest',
    eslint: true,
    prettier: true,
    commitlint: true,
    style: 'less',
    stylelint: true,
    devServer: 'docz'
};
exports.cli_components_vue = {
    project_type: 'component-vue',
    pkgtool: 'yarn',
    build: 'tsc',
    ts: true,
    testFrame: 'jest',
    eslint: true,
    prettier: true,
    commitlint: true,
    style: 'less',
    stylelint: true,
    devServer: 'storybook'
};
exports.cli_toolkit = {
    project_type: 'toolkit',
    pkgtool: 'yarn',
    build: 'rollup',
    ts: true,
    testFrame: 'mocha',
    eslint: true,
    prettier: true,
    commitlint: true,
    style: '',
    stylelint: false
};
exports.default = {
    cli_basic_react: exports.cli_basic_react,
    cli_standard_react: exports.cli_standard_react,
    cli_entire_react: exports.cli_entire_react,
    cli_pc_react: exports.cli_pc_react,
    cli_basic_vue: exports.cli_basic_vue,
    cli_standard_vue: exports.cli_standard_vue,
    cli_entire_vue: exports.cli_entire_vue,
    cli_ssr_react: exports.cli_ssr_react,
    cli_components_react: exports.cli_components_react,
    cli_components_vue: exports.cli_components_vue,
    cli_toolkit: exports.cli_toolkit
};
