// @omni-door/cli/lib/commands/commands.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var initial_1 = __importDefault(require("./initial"));
var dev_1 = __importDefault(require("./dev"));
var start_1 = __importDefault(require("./start"));
var new_1 = __importDefault(require("./new"));
var build_1 = __importDefault(require("./build"));
var release_1 = __importDefault(require("./release"));
var utils_1 = require("@omni-door/utils");
var initial_2 = require("./initial");
Object.defineProperty(exports, "initial", { enumerable: true, get: function () { return initial_2.default; } });
var dev_2 = require("./dev");
Object.defineProperty(exports, "dev", { enumerable: true, get: function () { return dev_2.default; } });
var start_2 = require("./start");
Object.defineProperty(exports, "start", { enumerable: true, get: function () { return start_2.default; } });
var new_2 = require("./new");
Object.defineProperty(exports, "newTpl", { enumerable: true, get: function () { return new_2.default; } });
var build_2 = require("./build");
Object.defineProperty(exports, "build", { enumerable: true, get: function () { return build_2.default; } });
var release_2 = require("./release");
Object.defineProperty(exports, "release", { enumerable: true, get: function () { return release_2.default; } });
var utils_2 = require("@omni-door/utils");
Object.defineProperty(exports, "setLogo", { enumerable: true, get: function () { return utils_2.setLogo; } });
Object.defineProperty(exports, "setBrand", { enumerable: true, get: function () { return utils_2.setBrand; } });
exports.default = {
    initial: initial_1.default,
    dev: dev_1.default,
    start: start_1.default,
    new: new_1.default,
    build: build_1.default,
    release: release_1.default,
    setLogo: utils_1.setLogo,
    setBrand: utils_1.setBrand
};
